tutorials/haskell/haskell-time-tutorial/tutorial.md

```diff
@@ -0,0 +1,481 @@
+to explain pattern synonyms in depth, instead it’s meant to be a practical
+introduction to this concept.
+
+GHC documentation describes pattern synonyms like:
+
+> “Pattern synonyms enable giving names to parametrized pattern schemes.
+> They can also be thought of as abstract constructors that don’t have a
+> bearing on data representation.”
+
+And probably we’re in the same place as in the beginning so I’ve preferred
+to introduce pattern synonyms with a real world example. We’re going to
+borrow a pattern that exists in the time library. The [MonthOfYear][month-pattern]
+pattern allows us to define months of the year not by a number but by its name.
+
+The `MonthOfYear` data type is defined like this:
+
+```haskell
+type MonthOfYear = Int
+```
+
+As you can see, it is only an alias for `Int`. We now are able to define
+months like `Int`s where January will correspond to 1 and so on. But
+fortunately, since GHC 7.8 we are able to use pattern synonyms and we
+could, instead of using raw `Int`s, write this down in a pattern style
+and have a more idiomatic way of defining months.
+
+```haskell
+pattern January :: MonthOfYear
+pattern January = 1
+
+pattern February :: MonthOfYear
+pattern February = 2
+
+pattern March :: MonthOfYear
+pattern March = 3
+.
+.
+.
+
+-- | The twelve 'MonthOfYear' patterns form a @COMPLETE@ set.
+pattern December :: MonthOf
+pattern December = 12
+
+{-# COMPLETE January, February, March, April, May, June, July, August, September, October, November, December #-}Year
+```
+
+If you’re curious about why the `{-# COMPLETE #-}` pragma is being
+used, see the [COMPLETE pragma documentation][complete-pragma] for more
+information.
+
+Now if we need to pass a `MonthOfYear` as a parameter to a function
+we could write something like:
+
+```haskell
+foo November -- or foo 11 --Both will be equivalent
+```
+
+We’ll use this pattern in a function that basically takes a month and
+a day and yields if the corresponding Date is a holiday.
+
+```haskell
+dateToHoliday :: MonthOfYear -> DayOfMonth -> String
+dateToHoliday January  _   = "Happy new year!"
+dateToHoliday November 1   = "Let's eat colada morada"
+dateToHoliday December 25  = "Merry Christmas ho ho ho"
+dateToHoliday _ _          = "Probably just a regular day"
+```
+
+(Now you're probably wondering what is “colada morada”? Haha no worries,
+it’s an [Ecuatorian tradition][colada-morada] and we eat this meal that
+is purple corn based with spices and fruits, pretty amazing to be honest.)
+
+Back to Haskell, I think it’s pretty clear what the `dateToHoliday`
+function does, the first clause matches all days of January, second
+one matches the All Saints holiday, the third one matches for Christmas
+and the last one it’s the fallback for all other days. I think using
+the `MonthOfYears` patterns makes it a lot more readable. Having something
+like:
+
+```haskell
+dateToHoliday 12 25  = "Merry Christmas jojojo"
```

Comment:
For showcasing the usage of a function I think is useful to use doctest syntax or something similar like a REPL output: 

```hs
> dateToHoliday 12 25
"Merry Christmas jojojo"
```